src/main/kotlin/com/terraformation/backend/tracking/model/PlantingZoneModel.kt

```diff
@@ -127,6 +137,174 @@ data class PlantingZoneModel(
+              as MultiPolygon
+        } else {
+          JTS.transform(boundary.difference(monitoringPlotAreas), toMeters) as MultiPolygon
+        }
+
+    fun rectangle(west: Double, south: Double, east: Double, north: Double): Polygon =
+        geometryFactory.createPolygon(
+            arrayOf(
+                Coordinate(west, south),
+                Coordinate(east, south),
+                Coordinate(east, north),
+                Coordinate(west, north),
+                Coordinate(west, south)))
+
+    /**
+     * Attempts to find an available square within a rectangular region of the zone. First tries to
+     * pick a few random points, and if none of them works, divides the region into four quadrants
+     * and searches them in area-weighted random order until it finds a match.
+     *
+     * The idea is that for a typical zone, the initial random probe will most likely work and we'll
+     * return a result quickly, but for an odd-shaped zone, we keep drilling down until we've done
+     * an exhaustive search for matches.
+     */
+    fun findInRegion(southwest: Coordinate, northeast: Coordinate): Polygon? {
+      val maxAttempts = 5
+
+      // Try a few times to find a random spot in the region. This will succeed most of the time.
+      for (attemptNumber in 1..maxAttempts) {
+        val squareWest =
+            Math.round(Random.nextDouble(southwest.x, northeast.x) / gridInterval) * gridInterval
+        val squareSouth =
+            Math.round(Random.nextDouble(southwest.y, northeast.y) / gridInterval) * gridInterval
+        val squareEast = squareWest + sizeMeters
+        val squareNorth = squareSouth + sizeMeters
+
+        val square = rectangle(squareWest, squareSouth, squareEast, squareNorth)
+
+        if (square.coveredBy(zoneGeometry)) {
+          return JTS.transform(square, toOriginalCrs) as Polygon
+        }
+      }
+
+      // No luck. This might be a sparse site map. Split the region into quartiles and search them
+      // in random order but with the ones that cover the most zone area more likely to come first.
+      val middleX = (southwest.x + northeast.x) / 2.0
+      val middleY = (southwest.y + northeast.y) / 2.0
+
+      // List of quadrant geometry and how much usable area the quadrant has.
+      val quadrants: MutableList<Pair<Polygon, Double>> =
+          listOf(
+                  rectangle(southwest.x, southwest.y, middleX, middleY),
+                  rectangle(middleX, southwest.y, northeast.x, middleY),
+                  rectangle(middleX, middleY, northeast.x, northeast.y),
+                  rectangle(southwest.x, middleY, middleX, northeast.y))
+              .map { quadrant -> quadrant to zoneGeometry.intersection(quadrant).area }
+              // Stop drilling down when we get to quadrants with 4 square meters of usable area
+              // since they're unlikely to round to different grid points than their parents.
+              .filter { it.second > 16.0 }
```

Comment:
Can you explain this a bit more? 16 seems like a magic number

Reply:
Good catch. This magic number made more sense in an earlier version of the code and I neglected to change it. Will address this.

Reply:
Reworked this a bit so the base case of the recursion is at the top of the function. Hopefully it's clearer what's going on now.